File: backend/src/llm/fallback_manager.py
from typing import Dict, List, Any, Optional
import asyncio
from datetime import datetime, timedelta
from collections import defaultdict
from .providers import ModelProviderFactory, ModelConfig, ModelResponse

class ModelFallbackManager:
    def __init__(self):
        self.health_checker = ModelHealthChecker()
        self.load_balancer = ModelLoadBalancer()
        self.fallback_chains = self._define_fallback_chains()
    
    def _define_fallback_chains(self) -> Dict[str, List[str]]:
        return {
            'gpt-4': ['gpt-4-turbo', 'claude-3-opus', 'gpt-3.5-turbo'],
            'claude-3-opus': ['claude-3-sonnet', 'gpt-4', 'claude-2.1'],
            'bedrock-claude': ['bedrock-titan', 'claude-3-opus', 'gpt-4'],
            'gpt-3.5-turbo': ['gpt-4', 'claude-3-haiku'],
            'claude-3-sonnet': ['claude-3-opus', 'gpt-4', 'gpt-3.5-turbo']
        }
    
    async def execute_with_fallback(
        self, 
        primary_model: str, 
        prompt: str, 
        options: Dict[str, Any]
    ) -> ModelResponse:
        fallback_chain = [primary_model] + self.fallback_chains.get(primary_model, [])
        last_error = None
        
        for model in fallback_chain:
            try:
                if not await self.health_checker.is_healthy(model):
                    continue
                
                if not await self.load_balancer.can_handle_request(model):
                    continue
                
                provider_name = model.split('-')[0]
                provider = ModelProviderFactory.create(
                    provider_name, 
                    ModelConfig(name=model, **options)
                )
                
                await provider.initialize()
                response = await provider.generate(prompt, options)
                
                await self._record_success(model)
                return response
                
            except Exception as e:
                last_error = e
                await self._record_failure(model, e)
                
                if not self._is_retryable_error(e):
                    raise
                
                continue
        
        raise Exception(f"All models in fallback chain failed. Last error: {last_error}")
    
    def _is_retryable_error(self, error: Exception) -> bool:
        retryable_errors = ['rate_limit', 'timeout', 'service_unavailable', 'internal_server_error']
        error_message = str(error).lower()
        return any(err in error_message for err in retryable_errors)
    
    async def _record_success(self, model: str) -> None:
        pass
    
    async def _record_failure(self, model: str, error: Exception) -> None:
        pass

class ModelHealthChecker:
    def __init__(self):
        self.health_status = defaultdict(lambda: {'healthy': True, 'last_check': datetime.now()})
    
    async def is_healthy(self, model: str) -> bool:
        status = self.health_status[model]
        
        # Check if we need to refresh health status
        if datetime.now() - status['last_check'] > timedelta(minutes=5):
            status['healthy'] = await self._check_model_health(model)
            status['last_check'] = datetime.now()
        
        return status['healthy']
    
    async def _check_model_health(self, model: str) -> bool:
        # Simple health check implementation
        return True

class ModelLoadBalancer:
    def __init__(self):
        self.request_counts = defaultdict(int)
        self.rate_limits = self._load_rate_limits()
    
    def _load_rate_limits(self) -> Dict[str, int]:
        return {
            'gpt-4': 500,
            'gpt-3.5-turbo': 3500,
            'claude-3-opus': 400,
            'claude-3-sonnet': 1000,
            'bedrock-claude': 200
        }
    
    async def can_handle_request(self, model: str) -> bool:
        current_count = self.request_counts[model]
        rate_limit = self.rate_limits.get(model, float('inf'))
        
        return current_count < rate_limit * 0.8  # 80% threshold

File: backend/src/llm/fallback-manager.ts
import Redis from 'ioredis';

interface FallbackChain {
  [key: string]: string[];
}

interface ModelHealth {
  model: string;
  isHealthy: boolean;
  lastCheck: Date;
  errorCount: number;
  responseTime: number;
}

export class ModelFallbackManager {
  private redis: Redis;
  private healthChecker: ModelHealthChecker;
  private loadBalancer: ModelLoadBalancer;
  private fallbackChains: FallbackChain;

  constructor() {
    this.redis = new Redis({
      host: process.env.REDIS_HOST || 'localhost',
      port: parseInt(process.env.REDIS_PORT || '6379'),
    });
    
    this.healthChecker = new ModelHealthChecker();
    this.loadBalancer = new ModelLoadBalancer(this.redis);
    this.fallbackChains = this.defineFallbackChains();
  }

  private defineFallbackChains(): FallbackChain {
    return {
      'gpt-4': ['gpt-4-turbo', 'claude-3-opus', 'gpt-3.5-turbo'],
      'claude-3-opus': ['claude-3-sonnet', 'gpt-4', 'claude-2.1'],
      'bedrock-claude': ['bedrock-titan', 'claude-3-opus', 'gpt-4'],
      'gpt-3.5-turbo': ['claude-3-haiku', 'gpt-4'],
      'claude-3-sonnet': ['claude-3-haiku', 'gpt-3.5-turbo'],
    };
  }

  async executeWithFallback(
    primaryModel: string,
    prompt: string,
    options: any = {}
  ): Promise<any> {
    const fallbackChain = [primaryModel, ...(this.fallbackChains[primaryModel] || [])];
    let lastError: Error | null = null;

    for (const model of fallbackChain) {
      try {
        if (!(await this.healthChecker.isHealthy(model))) {
          continue;
        }

        if (!(await this.loadBalancer.canHandleRequest(model))) {
          continue;
        }

        const response = await this.executeModel(model, prompt, options);
        await this.recordSuccess(model);
        return response;

      } catch (error) {
        lastError = error as Error;
        await this.recordFailure(model, error as Error);

        if (!this.isRetryableError(error as Error)) {
          throw error;
        }
      }
    }

    throw new Error(`All models failed. Last error: ${lastError?.message}`);
  }

  private async executeModel(model: string, prompt: string, options: any): Promise<any> {
    // Mock implementation - replace with actual model execution
    await new Promise(resolve => setTimeout(resolve, 100));
    return { content: `Response from ${model}: ${prompt.substring(0, 50)}...` };
  }

  private isRetryableError(error: Error): boolean {
    const retryableErrors = [
      'rate_limit', 'timeout', 'service_unavailable', 'internal_server_error'
    ];
    return retryableErrors.some(err => error.message.toLowerCase().includes(err));
  }

  private async recordSuccess(model: string): Promise<void> {
    await this.redis.hincrby('model:success', model, 1);
    await this.redis.hset('model:last_success', model, Date.now());
  }

  private async recordFailure(model: string, error: Error): Promise<void> {
    await this.redis.hincrby('model:failures', model, 1);
    await this.redis.hset('model:last_error', model, error.message);
  }
}

class ModelHealthChecker {
  private healthCache = new Map<string, ModelHealth>();
  private checkInterval = 60000; // 1분

  async isHealthy(model: string): Promise<boolean> {
    const cached = this.healthCache.get(model);
    
    if (cached && Date.now() - cached.lastCheck.getTime() < this.checkInterval) {
      return cached.isHealthy;
    }

    const health = await this.checkModelHealth(model);
    this.healthCache.set(model, health);
    return health.isHealthy;
  }

  private async checkModelHealth(model: string): Promise<ModelHealth> {
    const startTime = Date.now();
    
    try {
      // Mock health check - replace with actual implementation
      await new Promise(resolve => setTimeout(resolve, 50));
      
      return {
        model,
        isHealthy: true,
        lastCheck: new Date(),
        errorCount: 0,
        responseTime: Date.now() - startTime,
      };
    } catch (error) {
      return {
        model,
        isHealthy: false,
        lastCheck: new Date(),
        errorCount: 1,
        responseTime: Date.now() - startTime,
      };
    }
  }
}

class ModelLoadBalancer {
  private rateLimits: Map<string, number>;

  constructor(private redis: Redis) {
    this.rateLimits = new Map([
      ['gpt-4', 100],
      ['gpt-3.5-turbo', 1000],
      ['claude-3-opus', 50],
      ['claude-3-sonnet', 200],
      ['bedrock-claude', 500],
    ]);
  }

  async canHandleRequest(model: string): Promise<boolean> {
    const rateLimit = this.rateLimits.get(model) || 100;
    const currentCount = await this.getDistributedCount(model);
    
    return currentCount < rateLimit * 0.8; // 80% 임계값
  }

  private async getDistributedCount(model: string): Promise<number> {
    const key = `model_request_count:${model}`;
    const count = await this.redis.get(key);
    return parseInt(count || '0');
  }

  async incrementRequestCount(model: string): Promise<void> {
    const key = `model_request_count:${model}`;
    await this.redis.incr(key);
    await this.redis.expire(key, 60); // 1분 TTL
  }
}